File: cslacs/Visual Studio Templates/Item Templates/cs/Templates Source/EditableRootParent/EditableRootParent.cs
using System;
using Csla;
using System.Collections.Generic;

namespace $rootnamespace$
{
  [Serializable]
  public class $safeitemname$ : BusinessBase<$safeitemname$>
  {
    #region Business Methods

    // TODO: add your own fields, properties and methods

    // example with private backing field
    private static PropertyInfo<int> IdProperty =
      RegisterProperty(typeof($safeitemname$), new PropertyInfo<int>("Id"));
    private int _Id = IdProperty.DefaultValue;
    public int Id
    {
      get { return GetProperty(IdProperty, _Id); }
      set { SetProperty(IdProperty, ref _Id, value); }
    }

    // example with managed backing field
    private static PropertyInfo<string> NameProperty =
      RegisterProperty(typeof($safeitemname$), new PropertyInfo<string>("Name"));
    public string Name
    {
      get { return GetProperty(NameProperty); }
      set { SetProperty(NameProperty, value); }
    }

    private static PropertyInfo<$childlist$> ChildListProperty = 
      RegisterProperty<$childlist$>(typeof($safeitemname$), new PropertyInfo<$childlist$>("ChildList", "Child list"));
    public $childlist$ ChildList
    {
      get { return GetProperty<$childlist$>(ChildListProperty); }
    }

    private static PropertyInfo<$childitem$> ChildProperty = 
      RegisterProperty(typeof($safeitemname$), new PropertyInfo<$childitem$>("Child", "Child"));
    public $childitem$ Child
    {
      get { return GetProperty<$childitem$>(ChildProperty); }
    }
    #endregion

    #region Validation Rules

    protected override void AddBusinessRules()
    {
      // TODO: add validation rules
      //ValidationRules.AddRule(RuleMethod, "");
    }

    #endregion

    #region Authorization Rules

    protected override void AddAuthorizationRules()
    {
      // TODO: add authorization rules
      //AuthorizationRules.AllowWrite("Name", "Role");
    }

    private static void AddObjectAuthorizationRules()
    {
      // TODO: add authorization rules
      //AuthorizationRules.AllowEdit(typeof($safeitemname$), "Role");
    }

    #endregion

    #region Factory Methods

    public static $safeitemname$ New$safeitemname$()
    {
      return DataPortal.Create<$safeitemname$>();
    }

    public static $safeitemname$ Get$safeitemname$(int id)
    {
      return DataPortal.Fetch<$safeitemname$>(
        new SingleCriteria<$safeitemname$, int>(id));
    }

    public static void Delete$safeitemname$(int id)
    {
      DataPortal.Delete(new SingleCriteria<$safeitemname$, int>(id));
    }

    private $safeitemname$()
    { /* Require use of factory methods */ }

    #endregion

    #region Data Access

    [RunLocal]
    protected override void DataPortal_Create()
    {
      // TODO: load default values
      // omit this override if you have no defaults to set
      LoadProperty(ChildListProperty, $childlist$.New$childlist$());
      LoadProperty(ChildProperty, $childitem$.New$childitem$());
      base.DataPortal_Create();
    }

    private void DataPortal_Fetch(SingleCriteria<$safeitemname$, int> criteria)
    {
      // TODO: load values
      LoadProperty(ChildListProperty, $childlist$.Get$childlist$(null));
      LoadProperty(ChildProperty, $childitem$.Get$childitem$(null));
    }

    [Transactional(TransactionalTypes.TransactionScope)]
    protected override void DataPortal_Insert()
    {
      // TODO: insert values
      FieldManager.UpdateChildren(this);
    }

    [Transactional(TransactionalTypes.TransactionScope)]
    protected override void DataPortal_Update()
    {
      // TODO: update values
      FieldManager.UpdateChildren(this);
    }

    [Transactional(TransactionalTypes.TransactionScope)]
    protected override void DataPortal_DeleteSelf()
    {
      DataPortal_Delete(new SingleCriteria<$safeitemname$, int>(this.Id));
    }

    [Transactional(TransactionalTypes.TransactionScope)]
    private void DataPortal_Delete(SingleCriteria<$safeitemname$, int> criteria)
    {
      // TODO: delete values
      FieldManager.UpdateChildren(this);
    }

    #endregion
  }


  [Serializable]
  public class $childlist$ : BusinessListBase<$childlist$, $childitem$>
  {
    #region Factory Methods

    internal static $childlist$ New$childlist$()
    {
      return DataPortal.CreateChild<$childlist$>();
    }

    internal static $childlist$ Get$childlist$(
      object childData)
    {
      return DataPortal.FetchChild<$childlist$>(childData);
    }

    private $childlist$()
    { }

    #endregion

    #region Data Access

    private void Child_Fetch(object childData)
    {
      RaiseListChangedEvents = false;
      foreach (var child in (IList<object>)childData)
        this.Add($childitem$.Get$childitem$(child));
      RaiseListChangedEvents = true;
    }

    #endregion
  }


  [Serializable]
  public class $childitem$ : BusinessBase<$childitem$>
  {
    #region Business Methods

    // TODO: add your own fields, properties and methods

    // example with private backing field
    private static PropertyInfo<int> IdProperty =
      RegisterProperty(typeof($childitem$), new PropertyInfo<int>("Id"));
    private int _Id = IdProperty.DefaultValue;
    public int Id
    {
      get { return GetProperty(IdProperty, _Id); }
      set { SetProperty(IdProperty, ref _Id, value); }
    }

    // example with managed backing field
    private static PropertyInfo<string> NameProperty =
      RegisterProperty(typeof($childitem$), new PropertyInfo<string>("Name"));
    public string Name
    {
      get { return GetProperty(NameProperty); }
      set { SetProperty(NameProperty, value); }
    }

    #endregion

    #region Validation Rules

    protected override void AddBusinessRules()
    {
      // TODO: add validation rules
      //ValidationRules.AddRule(RuleMethod, "");
    }

    #endregion

    #region Authorization Rules

    protected override void AddAuthorizationRules()
    {
      // TODO: add authorization rules
      //AuthorizationRules.AllowWrite("Name", "Role");
    }

    private static void AddObjectAuthorizationRules()
    {
      // TODO: add authorization rules
      //AuthorizationRules.AllowEdit(typeof($childitem$), "Role");
    }

    #endregion

    #region Factory Methods

    internal static $childitem$ New$childitem$()
    {
      return DataPortal.CreateChild<$childitem$>();
    }

    internal static $childitem$ Get$childitem$(object childData)
    {
      return DataPortal.FetchChild<$childitem$>(childData);
    }

    private $childitem$()
    { /* Require use of factory methods */ }

    #endregion

    #region Data Access

    protected override void Child_Create()
    {
      // TODO: load default values
      // omit this override if you have no defaults to set
      base.Child_Create();
    }

    private void Child_Fetch(object childData)
    {
      // TODO: load values
    }

    private void Child_Insert(object parent)
    {
      // TODO: insert values
    }

    private void Child_Update(object parent)
    {
      // TODO: update values
    }

    private void Child_DeleteSelf(object parent)
    {
      // TODO: delete values
    }

    #endregion
  }
}

File: cslacs/Visual Studio Templates/Item Templates/vb/Templates Source/EditableRootParent/EditableRootParent.vb
Imports System
Imports System.Collections.Generic
Imports Csla

namespace $rootnamespace$

  <Serializable()> _
  Public Class $safeitemname$
    Inherits BusinessBase(Of $safeitemname$)

#Region "Business Methods"

    ' TODO: add your own fields, properties and methods 

    ' example with private backing field 
    Private Shared IdProperty As PropertyInfo(Of Integer) = RegisterProperty(New PropertyInfo(Of Integer)("Id"))
    Private _Id As Integer = IdProperty.DefaultValue
    Public Property Id() As Integer
      Get
        Return GetProperty(IdProperty, _Id)
      End Get
      Set(ByVal value As Integer)
        SetProperty(IdProperty, _Id, value)
      End Set
    End Property

    ' example with managed backing field 
    Private Shared NameProperty As PropertyInfo(Of String) = RegisterProperty(New PropertyInfo(Of String)("Name"))
    Public Property Name() As String
      Get
        Return GetProperty(NameProperty)
      End Get
      Set(ByVal value As String)
        SetProperty(NameProperty, value)
      End Set
    End Property

    Private Shared ChildListProperty As PropertyInfo(Of $childlist$) = RegisterProperty(Of $childlist$)(New PropertyInfo(Of $childlist$)("ChildList", "Child list"))
    Public ReadOnly Property ChildList() As $childlist$
      Get
        Return GetProperty(Of $childlist$)(ChildListProperty)
      End Get
    End Property

    Private Shared ChildProperty As PropertyInfo(Of $childitem$) = RegisterProperty(New PropertyInfo(Of $childitem$)("Child", "Child"))
    Public ReadOnly Property Child() As $childitem$
      Get
        Return GetProperty(Of $childitem$)(ChildProperty)
      End Get
    End Property
#End Region

#Region "Validation Rules"

    Protected Overloads Overrides Sub AddBusinessRules()
      ' TODO: add validation rules 
      'ValidationRules.AddRule(RuleMethod, ""); 
    End Sub

#End Region

#Region "Authorization Rules"

    Protected Overloads Overrides Sub AddAuthorizationRules()
      ' TODO: add authorization rules 
      AuthorizationRules.AllowWrite(NameProperty, "Role")
    End Sub

    Private Shared Sub AddObjectAuthorizationRules()
      ' TODO: add authorization rules 
      'AuthorizationRules.AllowEdit(GetType($childlist$), "Role")
    End Sub

#End Region

#Region "Factory Methods"

    Public Shared Function New$safeitemname$() As $safeitemname$
      Return DataPortal.Create(Of $safeitemname$)()
    End Function

    Public Shared Function Get$safeitemname$(ByVal id As Integer) As $safeitemname$
      Return DataPortal.Fetch(Of $safeitemname$)(New SingleCriteria(Of $safeitemname$, Integer)(id))
    End Function

    Public Shared Sub Delete$safeitemname$(ByVal id As Integer)
      DataPortal.Delete(New SingleCriteria(Of $safeitemname$, Integer)(id))
    End Sub

    ' Require use of factory methods 
    Private Sub New()
    End Sub

#End Region

#Region "Data Access"

    <RunLocal()> _
    Protected Overloads Overrides Sub DataPortal_Create()
      ' TODO: load default values 
      ' omit this override if you have no defaults to set 
      LoadProperty(ChildListProperty, $childlist$.New$childlist$())
      LoadProperty(ChildProperty, $childitem$.New$childitem$())
      MyBase.DataPortal_Create()
    End Sub

    Private Overloads Sub DataPortal_Fetch(ByVal criteria As SingleCriteria(Of $safeitemname$, Integer))
      ' TODO: load values 
      LoadProperty(ChildListProperty, $childlist$.Get$childlist$(Nothing))
      LoadProperty(ChildProperty, $childitem$.Get$childitem$(Nothing))
    End Sub

    <Transactional(TransactionalTypes.TransactionScope)> _
    Protected Overloads Overrides Sub DataPortal_Insert()
      ' TODO: insert values 
      FieldManager.UpdateChildren(Me)
    End Sub

    <Transactional(TransactionalTypes.TransactionScope)> _
    Protected Overloads Overrides Sub DataPortal_Update()
      ' TODO: update values 
      FieldManager.UpdateChildren(Me)
    End Sub

    <Transactional(TransactionalTypes.TransactionScope)> _
    Protected Overloads Overrides Sub DataPortal_DeleteSelf()
      DataPortal_Delete(New SingleCriteria(Of $safeitemname$, Integer)(Me.Id))
    End Sub

    <Transactional(TransactionalTypes.TransactionScope)> _
    Private Overloads Sub DataPortal_Delete(ByVal criteria As SingleCriteria(Of $safeitemname$, Integer))
      ' TODO: delete values 
      FieldManager.UpdateChildren(Me)
    End Sub

#End Region
  End Class



  <Serializable()> _
  Public Class $childlist$
    Inherits BusinessListBase(Of $childlist$, $childitem$)

#Region " Factory Methods "

    Friend Shared Function New$childlist$() As $safeitemname$
      Return DataPortal.CreateChild(Of $childlist$)()
    End Function

    Friend Shared Function Get$childlist$(ByVal childData As Object) As $childlist$
      Return DataPortal.FetchChild(Of $childlist$)(childData)
    End Function

    Private Sub New()
      'use factory methods
    End Sub

#End Region

#Region " Data Access "

    Private Sub Child_Fetch(ByVal childData As Object)
      RaiseListChangedEvents = False
      For Each child As Object In CType(childData, IList(Of Object))
        Me.Add($childitem$.Get$childitem$(child))
      Next
      RaiseListChangedEvents = True
    End Sub

#End Region

  End Class


  <Serializable()> _
  Public Class $childitem$
    Inherits BusinessBase(Of $childitem$)

#Region " Business Methods "

    ' TODO: add your own fields, properties and methods
    'example with private backing field
    Private Shared IdProperty As PropertyInfo(Of Integer) = RegisterProperty(New PropertyInfo(Of Integer)("Id", "Id"))
    Private _id As Integer = IdProperty.DefaultValue
    ''' <Summary>
    ''' Gets and sets the Id value.
    ''' </Summary>
    Public Property Id() As Integer
      Get
        Return GetProperty(IdProperty, _id)
      End Get
      Set(ByVal value As Integer)
        SetProperty(IdProperty, _id, value)
      End Set
    End Property

    'example with managed backing field
    Private Shared NameProperty As PropertyInfo(Of String) = RegisterProperty(New PropertyInfo(Of String)("Name", "Name"))
    ''' <Summary>
    ''' Gets and sets the Name value.
    ''' </Summary>
    Public Property Name() As String
      Get
        Return GetProperty(NameProperty)
      End Get
      Set(ByVal value As String)
        SetProperty(NameProperty, value)
      End Set
    End Property


#End Region

#Region " Validation Rules "

    Protected Overrides Sub AddBusinessRules()
      ' TODO: add validation rules
      'ValidationRules.AddRule(Nothing, "")
    End Sub

#End Region

#Region " Authorization Rules "

    Protected Overrides Sub AddAuthorizationRules()

      ' TODO: add authorization rules
      AuthorizationRules.AllowWrite(NameProperty, "")

    End Sub

    Private Shared Sub AddObjectAuthorizationRules()
      'TODO: add authorization rules
      AuthorizationRules.AllowEdit(GetType($childitem$), "Role")
    End Sub

#End Region

#Region " Factory Methods "

    Friend Shared Function New$childitem$() As $childitem$
      Return DataPortal.CreateChild(Of $safeitemname$)()
    End Function

    Friend Shared Function Get$childitem$(ByVal childData As Object) As $childitem$
      Return DataPortal.FetchChild(Of $safeitemname$)(childData)
    End Function

    Private Sub New()
      'Require use of factory methods
    End Sub

#End Region

#Region " Data Access "

    Protected Overrides Sub Child_Create()
      'TODO: load default values
      'omit this override if you have no defaults to set
      MyBase.Child_Create()
    End Sub

    Private Sub Child_Fetch(ByVal childData As Object)
      ' TODO: load values
    End Sub

    Private Sub Child_Insert(ByVal parent As Object)
      ' TODO: insert values
    End Sub

    Private Sub Child_Update(ByVal parent As Object)
      ' TODO: insert values
    End Sub

    Friend Sub Child_DeleteSelf(ByVal parent As Object)
      ' TODO: delete values
    End Sub

#End Region

  End Class

End Namespace